pkg/binary/envoy/fetch.go

```diff
@@ -15,128 +15,97 @@
+		if e != nil {
+			return "", e
 		}
 
-		err = fetchEnvoy(platformDirectory, binaryLocation)
-		if err != nil {
-			return "", err
+		downloadLocationURL, e := manifest.LocateBuild(ref, m)
+		if e != nil {
+			return "", e
 		}
-	case e != nil:
-		return "", fmt.Errorf("invalid Envoy binary at %q: %w", envoyPath, e)
-	default:
-		fmt.Printf("%v is already downloaded\n", ref)
-		if stat.Mode()&0111 == 0 {
-			return "", fmt.Errorf("envoy binary not executable: %s", envoyPath)
+
+		fmt.Fprintln(o.Out, "downloading", downloadLocationURL) //nolint
+		if e := untarEnvoy(platformPath, downloadLocationURL, o.Out); e != nil {
+			return "", e
 		}
+	} else {
+		fmt.Fprintln(o.Out, ref, "is already downloaded") //nolint
 	}
-
-	return envoyPath, nil
+	return verifyEnvoy(platformPath)
 }
 
-func fetchEnvoy(dst, src string) error {
-	tmpDir, err := ioutil.TempDir("", "getenvoy-")
+func verifyEnvoy(platformPath string) (string, error) {
+	envoyPath := filepath.Join(platformPath, binEnvoy)
+	stat, err := os.Stat(envoyPath)
 	if err != nil {
-		return err
-	}
-	defer os.RemoveAll(tmpDir) //nolint
-	tarball, err := doDownload(tmpDir, src)
-	if err != nil {
-		return fmt.Errorf("unable to fetch envoy from %v: %w", src, err)
+		return "", err
 	}
-	err = extractEnvoy(dst, tarball)
-	if err != nil {
-		return fmt.Errorf("unable to extract envoy to %v: %w", dst, err)
+	if stat.Mode()&0111 == 0 {
+		return "", fmt.Errorf("envoy binary not executable at %q", envoyPath)
 	}
-	return nil
+	return envoyPath, nil
 }
 
-func doDownload(dst, src string) (string, error) {
-	// #nosec -> src can be anywhere by design
-	resp, err := transport.Get(src)
+func untarEnvoy(dst, url string, out io.Writer) error { // dst, src order like io.Copy
+	// #nosec -> url can be anywhere by design
+	resp, err := transport.Get(url)
 	if err != nil {
-		return "", err
+		return err
 	}
 	defer resp.Body.Close() //nolint
 
 	if resp.StatusCode != http.StatusOK {
-		return "", fmt.Errorf("received %v status code from %q", resp.StatusCode, src)
+		return fmt.Errorf("received %v status code from %s", resp.StatusCode, url)
 	}
 
-	tarball := filepath.Join(dst, "envoy.tar"+filepath.Ext(src))
-	// #nosec -> dst can be anywhere by design
-	f, err := os.OpenFile(tarball, os.O_CREATE|os.O_WRONLY, 0600)
-	if err != nil {
-		return "", err
-	}
-	defer f.Close() //nolint
-
-	bar := progressbar.NewOptions64(resp.ContentLength, progressbar.OptionSetDescription("[Fetching Envoy]"))
-	out := io.MultiWriter(f, bar)
-	_, err = io.Copy(out, resp.Body)
-	fmt.Println("") // append a newline to progressbar output
-	return tarball, err
-}
+	// Ensure there's a progress bar while extraction is taking place
+	bar := progressbar.NewOptions64(
+		resp.ContentLength,
+		progressbar.OptionSetWriter(out),
+		progressbar.OptionOnCompletion(func() {
+			fmt.Fprint(out, "\n")
+		}),
+	)
+	defer bar.Close() //nolint
+	src := progressbar.NewReader(resp.Body, bar)
 
-func extractEnvoy(dst, tarball string) error {
```

Comment:
we now extract the whole directory, which makes this forwards compatible with other inclusions such as tools